relayer/chains/substrate/event_parser.go

```diff
@@ -1,29 +1,501 @@
-	"github.com/cosmos/relayer/v2/relayer/processor"
 	"github.com/cosmos/relayer/v2/relayer/provider"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
 )
 
-// ibcMessagesFromTransaction parses all events within a transaction to find IBC messages
-func (scp *SubstrateChainProcessor) handleIBCMessagesFromEvents(ibcEvents rpcclienttypes.IBCEventsQueryResult, height uint64, c processor.IBCMessagesCache) error {
-	//TODO implement me
-	panic("implement me")
+// ibcMessage is the type used for parsing all possible properties of IBC messages
+type ibcMessage struct {
+	eventType string
+	info      ibcMessageInfo
+}
+
+type ibcMessageInfo interface {
+	parseAttrs(log *zap.Logger, attrs interface{})
+	MarshalLogObject(enc zapcore.ObjectEncoder) error
+}
+
+// alias for the interface map
+type ibcEventQueryItem (map[string]interface{})
+
+// substrate ibc events endpoint returns a list of events
+// relayted to different transaction, so we need
+// to define a unique key
+type ibcPacketKey struct {
+	sequence   uint64
+	srcChannel string
+	srcPort    string
+	dstChannel string
+	dstPort    string
+}
+
+// ibcMessagesFromEvents parses all events of a certain height to find IBC messages
+func (scp *SubstrateChainProcessor) ibcMessagesFromEvents(
+	ibcEvents rpcclienttypes.IBCEventsQueryResult,
+	height uint64,
+) (messages []ibcMessage) {
+
+	packetAccumulator := make(map[ibcPacketKey]*packetInfo)
+	for i := 0; i < len(ibcEvents); i++ {
+
+		info, eventType := scp.parseEvent(ibcEvents[i], height, packetAccumulator)
+		if info == nil {
+			// Not an IBC message, don't need to log here
+			continue
+		}
+
+		// update messages
+		messages = append(messages, ibcMessage{
+			eventType: eventType,
+			info:      info,
+		})
+	}
+
+	return messages
+}
+
+func (scp *SubstrateChainProcessor) parseEvent(
+	event ibcEventQueryItem,
+	height uint64,
+	packetAccumulator map[ibcPacketKey]*packetInfo,
+) (info ibcMessageInfo, eventType string) {
+
+	// there is one itm in the event type map
+	// so this loop iterates just one iteration
+	for eType, data := range event {
+
+		switch eType {
+
+		case CreateClient, UpgradeClient, ClientMisbehaviour:
+
+			cl := new(clientInfo)
+			cl.parseAttrs(scp.log, data)
+			info = cl
+
+			eventType = intoIBCEventType(eType)
+
+		case UpdateClient:
+
+			clu := new(clientUpdateInfo)
+			clu.parseAttrs(scp.log, data)
+			info = clu
+
+			eventType = intoIBCEventType(eType)
+
+		case SendPacket, AcknowledgePacket, TimeoutPacket, TimeoutOnClosePacket:
+
+			pkt := new(packetInfo)
+			pkt.parseAttrs(scp.log, data)
+			info = pkt
+
+			eventType = intoIBCEventType(eType)
+
+		case ReceivePacket, WriteAcknowledgement:
+
+			accumKey := genAccumKey(data)
+
+			_, exists := packetAccumulator[accumKey]
+			if !exists {
+				packetAccumulator[accumKey] = &packetInfo{Height: height}
+			}
+
+			packetAccumulator[accumKey].parseAttrs(scp.log, data)
+
+			info = packetAccumulator[accumKey]
```

Comment:
We don't want to set info for this case.

Reply:
Done